File: src/main/java/alfheim/common/block/tile/sub/SubTileLightning.java
package alfheim.common.block.tile.sub;

import java.util.ArrayList;
import java.util.List;

import alexsocol.asjlib.ASJUtilities;
import alfheim.api.block.tile.SubTileEntity;
import alfheim.api.block.tile.SubTileEntity.EnumAnomalityRarity;
import alfheim.common.core.handler.CardinalSystem.KnowledgeSystem;
import alfheim.common.core.handler.CardinalSystem.KnowledgeSystem.Knowledge;
import alfheim.common.core.registry.AlfheimBlocks;
import alfheim.common.core.util.AlfheimConfig;
import alfheim.common.core.util.DamageSourceSpell;
import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.effect.EntityLightningBolt;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.item.ItemStack;
import net.minecraft.util.AxisAlignedBB;
import net.minecraft.util.MathHelper;
import net.minecraft.world.World;
import vazkii.botania.common.Botania;
import vazkii.botania.common.block.ModBlocks;
import vazkii.botania.common.block.tile.TileRuneAltar;
import vazkii.botania.common.core.helper.Vector3;
import vazkii.botania.common.item.ModItems;

public class SubTileLightning extends SubTileEntity {
	
	public static final double radius = 12;
	Vector3 vt = new Vector3();
	Vector3 ve = new Vector3();
	
	@Override
	public void update() {
		if (inWG()) return;
		
		vt.set(x() + 0.5, y() + 0.5, z() + 0.5);
		
		if (worldObj().rand.nextInt(6000) == 0) {
			double x = x() + Math.random() * 10 - 5, z = z() + Math.random() * 10 - 5;
			worldObj().addWeatherEffect(new EntityLightningBolt(worldObj(), x, worldObj().getTopSolidOrLiquidBlock(MathHelper.floor_double(x), MathHelper.floor_double(z)), z));
			return;
		}
		
		if (AlfheimConfig.lightningsSpeed > 0 && ticks % AlfheimConfig.lightningsSpeed == 0) {
			ve.set(Math.random() - 0.5, Math.random() - 0.5, Math.random() - 0.5).normalize();
			vt.add(ve.x / 2.25, ve.y / 2.25, ve.z / 2.25);
			ve.multiply(1.5).add(x() + 0.5, y() + 0.5, z() + 0.5);
			Botania.proxy.lightningFX(worldObj(), vt, ve, 50, worldObj().rand.nextLong(), 0, 0xFF0000);
		}
	}
	
	@Override
	public boolean onActivated(ItemStack stack, EntityPlayer player, World world, int x, int y, int z) {
		if (player instanceof EntityPlayerMP && stack != null && stack.getItem() == ModItems.manaResource && stack.stackSize > 0 && stack.getItemDamage() == 5 && world.getBlock(x, y - 1, z) == ModBlocks.pylon && world.getBlockMetadata(x, y - 1, z) == 2) {
			--stack.stackSize;
			world.setBlock(x, y - 1, z, AlfheimBlocks.alfheimPylons, 2, 3);
			
			world.playSoundEffect(x, y, z, "botania:runeAltarStart", 1F, 1F);
			
			for (int i = 0; i < 8; i++) {
				ve.set(Math.random() - 0.5, Math.random() - 0.5, Math.random() - 0.5).normalize();
				vt.add(ve.x / 2.25, ve.y / 2.25, ve.z / 2.25);
				ve.multiply(1.5).add(x() + 0.5, y() + 0.5, z() + 0.5);
				Botania.proxy.lightningFX(worldObj(), vt, ve, 50, worldObj().rand.nextLong(), 0, 0xFF0000);
			}
			
			KnowledgeSystem.learn((EntityPlayerMP) player, Knowledge.PYLONS);
		}
		return false;
	}
	
	@Override
	public List<Object> getTargets() {
		if (inWG()) return EMPTY_LIST;
		
		f: if (ticks % 50 == 0) {
			EntityLivingBase e =  findNearestVulnerableEntity(radius);
			if (e == null) break f;
			
			worldObj().playSoundEffect(x(), y(), z(), "botania:runeAltarCraft", 1F, 1F);
			
			List l = new ArrayList<Object>();
			l.add(e);
			return l;
		}
		return EMPTY_LIST;
	}
	
	@Override
	public void performEffect(Object target) {
		if (ticks % 25 != 0) return;
		if (target == null || !(target instanceof EntityLivingBase)) return;
		if (target instanceof EntityPlayer && ((EntityPlayer) target).capabilities.disableDamage) return;
		
		EntityLivingBase entity = (EntityLivingBase) target;
		
		entity.attackEntityFrom(DamageSourceSpell.corruption, (float) Math.min((Math.random() * 2 + 3) / vt.set(x() + 0.5, y() + 0.5, z() + 0.5).add(-entity.posX, -entity.posY, -entity.posZ).mag() / 2, 2.5) * 4F);
		
		vt.set(x() + 0.5, y() + 0.5, z() + 0.5);
		ve.set(entity.posX, entity.posY, entity.posZ).normalize();
		vt.add(ve.x / 2, ve.y / 2, ve.z / 2);
		ve.set(entity.posX, entity.posY, entity.posZ);
		
		Botania.proxy.lightningFX(worldObj(), vt, ve, 1, worldObj().rand.nextLong(), 0, 0xFF0000);
	}
	
	@Override
	public int typeBits() {
		return HEALTH;
	}
	
	@Override
	public int getStrip() {
		return 1;
	}
	
	@Override
	public EnumAnomalityRarity getRarity() {
		return EnumAnomalityRarity.COMMON;
	}
}

File: src/main/java/alfheim/common/block/tile/sub/anomaly/SubTileLightning.kt
package alfheim.common.block.tile.sub.anomaly

import alexsocol.asjlib.*
import alfheim.api.block.tile.SubTileAnomalyBase
import alfheim.common.block.AlfheimBlocks
import alfheim.common.core.handler.AlfheimConfigHandler
import alfheim.common.core.handler.CardinalSystem.KnowledgeSystem
import alfheim.common.core.handler.CardinalSystem.KnowledgeSystem.Knowledge
import alfheim.common.core.util.DamageSourceSpell
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.effect.EntityLightningBolt
import net.minecraft.entity.player.*
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.world.World
import net.minecraftforge.common.util.ForgeDirection
import vazkii.botania.common.Botania
import vazkii.botania.common.block.ModBlocks
import vazkii.botania.common.core.helper.Vector3
import vazkii.botania.common.item.ModItems
import java.util.*
import kotlin.math.min
import alexsocol.asjlib.math.Vector3 as ASJVec3

class SubTileLightning: SubTileAnomalyBase() {
	
	internal val vt = Vector3()
	internal val ve = Vector3()
	
	override val targets: List<Any>
		get() {
			if (inWG()) return EMPTY_LIST
			
			run {
				if (ticks % 50 == 0) {
					val e = findNearestVulnerableEntity(radius) ?: return@run
					
					worldObj.playSoundEffect(x().D, y().D, z().D, "botania:runeAltarCraft", 1f, 1f)
					
					val l = ArrayList<Any>()
					l.add(e)
					return l
				}
			}
			return EMPTY_LIST
		}
	
	override val strip: Int
		get() = 1
	
	override val rarity: EnumAnomalityRarity
		get() = EnumAnomalityRarity.COMMON
	
	var transfer = 0
	var side = ForgeDirection.UNKNOWN
	
	public override fun update() {
		if (inWG()) return
		
		vt.set(x() + 0.5, y() + 0.5, z() + 0.5)
		
		if (transfer > 0) {
			--transfer
			
			if (transfer % 10 == 0) {
				
				val yOff = if (transfer % 20 == 0) 1.5 else 0.0
				
				vt.set(x() + 0.5, y() + 0.5, z() + 0.5)
				ve.set(x() + side.offsetX + 0.5, y() + side.offsetY + yOff, z() + side.offsetZ + 0.5).normalize()
				vt.add(ve.x / 2, ve.y / 2, ve.z / 2)
				ve.set(x() + side.offsetX + 0.5, y() + side.offsetY + yOff, z() + side.offsetZ + 0.5)
				
				Botania.proxy.lightningFX(worldObj, vt, ve, 1f, worldObj.rand.nextLong(), 0, 0xFF0000)
			}
			
			if (side === ForgeDirection.UNKNOWN || worldObj.getBlock(x() + side.offsetX, y() + side.offsetY, z() + side.offsetZ) !== ModBlocks.pylon || worldObj.getBlockMetadata(x() + side.offsetX, y() + side.offsetY, z() + side.offsetZ) != 2) {
				transfer = 0
				worldObj.newExplosion(null, x().D, y().D, z().D, 10f, false, true)
				return
			}
			
			if (transfer == 1) {
				worldObj.playSoundEffect(x().D, y().D, z().D, "botania:runeAltarCraft", 1f, 1f)
				
				for (i in 0..7) {
					ve.set(Math.random() - 0.5, Math.random() - 0.5, Math.random() - 0.5).normalize()
					vt.add(ve.x / 2.25, ve.y / 2.25, ve.z / 2.25)
					ve.multiply(1.5).add(x() + side.offsetX + 0.5, y() + side.offsetY + 0.5, z() + side.offsetZ + 0.5)
					Botania.proxy.lightningFX(worldObj, vt, ve, 50f, worldObj.rand.nextLong(), 0, 0xFF0000)
				}
				
				worldObj.setBlock(x() + side.offsetX, y() + side.offsetY, z() + side.offsetZ, AlfheimBlocks.alfheimPylon, 2, 3)
				worldObj.setBlockToAir(x(), y(), z())
			}
			
			return
		}
		
		if (worldObj.rand.nextInt(6000) == 0) {
			val x = x() + Math.random() * 10 - 5
			val z = z() + Math.random() * 10 - 5
			worldObj.addWeatherEffect(EntityLightningBolt(worldObj, x, worldObj.getTopSolidOrLiquidBlock(x.mfloor(), z.mfloor()).D, z))
			return
		}
		
		if (ASJUtilities.isClient) {
			if (superTile != null && ASJVec3.entityTileDistance(mc.thePlayer, superTile!!) > 32) return
			
			if (AlfheimConfigHandler.lightningsSpeed > 0 && ticks % AlfheimConfigHandler.lightningsSpeed == 0) {
				ve.set(Math.random() - 0.5, Math.random() - 0.5, Math.random() - 0.5).normalize()
				vt.add(ve.x / 2.25, ve.y / 2.25, ve.z / 2.25)
				ve.multiply(1.5).add(x() + 0.5, y() + 0.5, z() + 0.5)
				Botania.proxy.lightningFX(worldObj, vt, ve, 50f, worldObj.rand.nextLong(), 0, 0xFF0000)
			}
		}
	}
	
	override fun onActivated(stack: ItemStack?, player: EntityPlayer, world: World, x: Int, y: Int, z: Int): Boolean {
		stack ?: return false
		if (stack.item === ModItems.manaResource && stack.stackSize > 0 && stack.meta == 5) {
			--stack.stackSize
			
			for (d in ForgeDirection.VALID_DIRECTIONS) {
				if (world.getBlock(x + d.offsetX, y + d.offsetY, z + d.offsetZ) === ModBlocks.pylon && world.getBlockMetadata(x + d.offsetX, y + d.offsetY, z + d.offsetZ) == 2) {
					transfer = 200
					side = d
					break
				}
			}
			
			world.playSoundEffect(x.D, y.D, z.D, "botania:runeAltarStart", 1f, 1f)
			
			for (i in 0..7) {
				ve.set(Math.random() - 0.5, Math.random() - 0.5, Math.random() - 0.5).normalize()
				vt.add(ve.x / 2.25, ve.y / 2.25, ve.z / 2.25)
				ve.multiply(1.5).add(x() + 0.5, y() + 0.5, z() + 0.5)
				Botania.proxy.lightningFX(worldObj, vt, ve, 50f, worldObj.rand.nextLong(), 0, 0xFF0000)
			}
			
			if (player is EntityPlayerMP)
				KnowledgeSystem.learn(player, Knowledge.PYLONS)
		}
		return false
	}
	
	override fun performEffect(target: Any) {
		if (transfer > 0) return
		if (ticks % 25 != 0) return
		if (target !is EntityLivingBase) return
		if (target is EntityPlayer && target.capabilities.disableDamage) return
		
		target.attackEntityFrom(DamageSourceSpell.anomaly, min((Math.random() * 2 + 3) / vt.set(x() + 0.5, y() + 0.5, z() + 0.5).add(-target.posX, -target.posY, -target.posZ).mag() / 2.0, 2.5).F * 4f)
		
		vt.set(x() + 0.5, y() + 0.5, z() + 0.5)
		ve.set(target.posX, target.posY, target.posZ).normalize()
		vt.add(ve.x / 2, ve.y / 2, ve.z / 2)
		ve.set(target.posX, target.posY, target.posZ)
		
		Botania.proxy.lightningFX(worldObj, vt, ve, 1f, worldObj.rand.nextLong(), 0, 0xFF0000)
	}
	
	override fun typeBits() = HEALTH
	
	override fun writeCustomNBT(cmp: NBTTagCompound) {
		super.writeCustomNBT(cmp)
		
		cmp.setInteger(TAG_SIDE, side.ordinal)
		if (side != ForgeDirection.UNKNOWN)
			cmp.setInteger(TAG_TRANSFER, transfer)
	}
	
	override fun readCustomNBT(cmp: NBTTagCompound) {
		super.readCustomNBT(cmp)
		
		side = ForgeDirection.getOrientation(cmp.getInteger(TAG_SIDE))
		if (side != ForgeDirection.UNKNOWN)
			transfer = cmp.getInteger(TAG_TRANSFER)
	}
	
	companion object {
		
		const val radius = 12.0
		
		const val TAG_TRANSFER = "transfer"
		const val TAG_SIDE = "side"
	}
}